Community Events. Display Community Events in Calypso.

andfinally opened the pull request:
HACK week WIP.

andfinally commented on client/reader/sidebar/index.jsx:
```diff
@@ -207,7 +213,27 @@ export class ReaderSidebar extends React.Component {
+										href="/read/community-events"
+										onClick={ this.handleReaderSidebarCommunityEventsClicked }
+									>
+										<Gridicon icon="reader" size={ 24 } />
```
Good thought Spen, I'm now using that one.

andfinally commented on client/state/community-events/actions.js:
```diff
@@ -0,0 +1,66 @@
+			type: COMMUNITY_EVENTS_REQUEST,
+		} );
+
+		return new Promise( resolve => {
```
Thanks Spen, this is a good improvement! I've tweaked `getCommunityEvents` to return the promise from `wpcom.req.get` and updated this function as you suggested.

michelleweber commented:
Looks good. 

On the "sorry" page could we tell them where the closest thing is?  Or offer them something else -- a link to wp.tv, a "Psst, you could start one" note with a link, something like that?

andfinally commented:
I've changed to the serif font to fit in better with the Reader content.

<img width="1024" alt="community-events-serif" src="https://user-images.githubusercontent.com/1647564/45101542-9cf3df80-b123-11e8-82ae-f4c32938492c.png">

<img width="1025" alt="community-events-sorry-serif" src="https://user-images.githubusercontent.com/1647564/45101552-a2e9c080-b123-11e8-8015-1382baabf1d5.png">

spen commented:
Design-wise I can see us making this very pretty over time. I'm not sure but I would guess it'd be possible to add an image to each event too, to display in the same, or similar, way to how blog post with featured images appear - making the event items more engaging :)

As we're behind the 'reader/community-events' flag here I wouldn't stress too much on visuals for now though - we can always iterate on this after more a12s have seen and tested it in its current state :)

andfinally commented on client/reader/community-events/index.jsx:
```diff
@@ -0,0 +1,164 @@
+		return events.map( ( event, index ) => {
+			const {
+				title = '',
+				meetup = '',
```
I thought I should follow the names used by the wordpress.org API here - it's the name of the regular meetup of which this is one event, so quite hard to name descriptively.

andfinally commented on client/reader/community-events/index.jsx:
```diff
@@ -0,0 +1,164 @@
+			} = event;
+
+			return (
+				<CompactCard key={ index } className="community-events__event">
```
TIL!

andfinally commented:
Thanks a lot for all the brilliant suggestions @spen! I've implemented all of them except one minor one.

* Added selectors, with unit tests for `getCommunityEvents` selector.
* Renamed `renderNoEvents` and some classes to be more informative.
* Pulled event out into a separate component, passing URL to tracking when user clicks the external link.
* Pulled empty events into a separate component, simplifying the conditions in the main `render` method.
* Using a better key on event components.
* Changed URL for meetups in empty component.

It's hard to figure out how to rename the `meetup` property - to convey the proper meaning it would need to be something like `meetupWhichIsHoldingEvent`, which seems a bit clunky to me.

a8ci18n commented on client/reader/community-events/event.jsx:
```diff
@@ -0,0 +1,71 @@
+							className="community-events__event-link"
+							onClick={ this.recordExternalEventLinkClick }
+						>
+							{ this.props.translate( 'Event details' ) }
```
Hi! I've found a possible matching string that has already been translated **15** times:
`translate( 'Event Details' )` *ES Score: 17*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

andfinally replied:
Good idea, thanks @jblz!

mattsherman commented:
I added a few tasks to this issue based on discussion in p8F0Ux-1do-p2

andfinally commented:
I've hopefully addressed the caching issue on the endpoint in https://code.a8c.com/D18510 - it uses a similar approach to the one we're using in `wp_localize_community_events`. Also added the flag to Horizon.

spen commented on client/state/community-events/actions.js:
```diff
@@ -0,0 +1,62 @@
+			.then( data => {
+				dispatch( communityEventsReceive( data ) );
+
+				return data;
```
I'd argue that it's good practice to always pass on the data but you are right - there's no other `.then` consuming that data.

I had a related discussion on this at https://github.com/Automattic/wp-calypso/pull/27395#issuecomment-423991731 last week that might be interesting.

jblz replied:
Pass 'em all and let the GC sort it out! :)

andfinally replied:
I was in two minds about this - I do think it's usually not a bad idea to return something. But when tweaking this to debug something else I had a few minutes of confusion about whether it's the returned data or the dispatched data that is rendered. So I've removed the return to make it clearer what the function is doing.

spen commented on client/reader/community-events/placeholder.jsx:
```diff
@@ -0,0 +1,36 @@
+		return (
+			<CompactCard key={ 0 } className="community-events__event-placeholder">
+				<CardHeading tagName="h2" size={ 21 }>
+					Loading interesting events...
```
I spotted these too, but they'll not be readable. Instead they'll appear as flashing placeholder blocks.

andfinally replied:
That's right, the text is invisible, and only here to take the right amount of space.

jblz commented on client/reader/community-events/index.jsx:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	renderLoading() {
+		const count = get( this.props, 'events.length', 2 );
```
Good catch, @spen -- come to think of it, I did notice a flash of an empty-looking placeholder screen.

andfinally replied:
TIL! Very good point Spen, thanks.

andfinally commented on client/reader/community-events/empty.jsx:
```diff
@@ -0,0 +1,42 @@
+							'world of WordPress, or {{link2}}find a meetup{{/link2}} near you.',
+						{
+							components: {
+								link1: <ExternalLink href="https://wordpress.tv/" icon />,
```
Thanks Jeff, done!

andfinally commented on client/reader/community-events/placeholder.jsx:
```diff
@@ -0,0 +1,36 @@
+class EventPlaceholder extends React.PureComponent {
+	render() {
+		return (
+			<CompactCard key={ 0 } className="community-events__event-placeholder">
```
Argh, it is not useful, thanks!

andfinally commented on client/reader/community-events/placeholder.jsx:
```diff
@@ -0,0 +1,36 @@
+						<div className="community-events__event-location">Event location</div>
+					</div>
+					<ExternalLink icon={ true } className="community-events__event-link">
+						Event details
```
As above...

andfinally commented on client/state/selectors/get-community-events-error.js:
```diff
@@ -0,0 +1,16 @@
+ * @param {Object} state Global state tree
+ * @returns {(string|null)} Error message, or null
+ */
+export default function getCommunityEvents( state ) {
```
Another good spot, this was a mistake, thanks!

andfinally commented:
I've hopefully now addressed Jeff and Spen's latest points: I've added the config setting to wpcalypso, and fixed an issue with the `CommunityEvent` key prop - we were passing it a `location` object instead of a string.
